### PR TITLE
Fix tbaa annotation on union selector bytes inside of structs 

### LINKED ISSUE #54599: PkgId objects change by printing and returning from function
Running the following prints the UUID of Aqua as expected (on 1.10.3, 1.11.0-beta1, and b54dce216bf (few days old master)):
```julia
using Pkg
Pkg.activate(; temp=true)
Pkg.add(["Aqua"])
Pkg.instantiate()

function foo()
    pkgid = Base.identify_package("Aqua")
    return pkgid
end

println(foo().uuid)
```

However, if we move the printing into the function, makes the code behave differently in different julia versions:
```julia
using Pkg
Pkg.activate(; temp=true)
Pkg.add(["Aqua"])
Pkg.instantiate()

function foo()
    pkgid = Base.identify_package("Aqua")
    println(pkgid.uuid)
end

foo()
```
output 1.10.3:
```
4c88cf16-eb10-579e-8560-4a9242c79595
```
output 1.11.0-beta1:
```
ERROR: LoadError: type Nothing has no field uuid
Stacktrace:
 [1] getproperty
   @ ./Base.jl:49 [inlined]
 [2] foo()
   @ Main ~/code/julia/julia/foo.jl:9
 [3] top-level scope
   @ ~/code/julia/julia/foo.jl:13
```
output b54dce216bf (few days old master):
```
ERROR: LoadError: type Nothing has no field uuid
Stacktrace:
 [1] getproperty
   @ ./Base.jl:49 [inlined]
 [2] foo()
   @ Main ~/code/julia/julia/foo.jl:9
 [3] top-level scope
   @ ~/code/julia/julia/foo.jl:13
 [4] include
   @ ./Base.jl:559 [inlined]
 [5] exec_options(opts::Base.JLOptions)
   @ Base ./client.jl:325
 [6] _start()
   @ Base ./client.jl:533
```

If we add another print in there, the LoadError transforms into a segfault:
```julia
using Pkg
Pkg.activate(; temp=true)
Pkg.add(["Aqua"])
Pkg.instantiate()

function foo()
    pkgid = Base.identify_package("Aqua")
    println(pkgid)
    println(pkgid.uuid)
end

foo()
```
output 1.10.3:
```
Aqua [4c88cf16-eb10-579e-8560-4a9242c79595]
4c88cf16-eb10-579e-8560-4a9242c79595
```
output 1.11.0-beta1:
```
Base.PkgId(Base.UUID("4c88cf16-eb10-579e-8560-4a9242c79595"), "Aqua")

[925569] signal 11 (128): Segmentation fault
in expression starting at /home/lgoe/code/julia/julia/foo.jl:13
foo at /home/lgoe/code/julia/julia/foo.jl:6
unknown function (ip: 0x7417e9b3158f)
unknown function (ip: (nil))
Allocations: 6884270 (Pool: 6882848; Big: 1422); GC: 8
[1]    925569 segmentation fault (core dumped)  julia +1.11 foo.jl
```
output b54dce216bf (few days old master):
```
Base.PkgId(Base.UUID("4c88cf16-eb10-579e-8560-4a9242c79595"), "Aqua")

[925819] signal 11 (128): Segmentation fault
in expression starting at /home/lgoe/code/julia/julia/foo.jl:13
foo at /home/lgoe/code/julia/julia/foo.jl:6
unknown function (ip: 0x7d7afbb2adff)
unknown function (ip: (nil))
Allocations: 4446232 (Pool: 4444892; Big: 1340); GC: 6
[1]    925819 segmentation fault (core dumped)  julia +nightly foo.jl
```

some device information:
```
Platform Info:
  OS: Linux (x86_64-linux-gnu)
  CPU: 8 × Intel(R) Core(TM) i7-10510U CPU @ 1.80GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-15.0.7 (ORCJIT, skylake)
```
all three julia versions are installed via juliaup

cc @fingolfin

### PR DESCRIPTION
**Pull request by gbaraldi**
We currently cause a alias analysis contradiction by saying that the unionselbytes are on the stack, even if they are on a struct. LLVM is then able to figure out that we giving it a impossible alias situation (the object doesn't alias itself) and triggers UB.

https://godbolt.org/z/ssEKMzsPf

We may want to do a benchmarks run on this to see if anything too critical hasn't regressed.

Fixes https://github.com/JuliaLang/julia/issues/54599

### CODE REVIEW

File: src/cgutils.cpp [hunk 1]
```diff
@@ -2780,7 +2780,8 @@ static MDNode *best_field_tbaa(jl_codectx_t &ctx, const jl_cgval_t &strct, jl_da
         }
     }
     if (strct.V && jl_field_isconst(jt, idx) && isLoadFromConstGV(strct.V))
-        return ctx.tbaa().tbaa_const;
+        return ctx.tbaa().tbaa_const; //TODO: it seems odd to have a field with a tbaa that doesn't alias it's containing struct's tbaa
```
Comment (fingolfin) [inline src/cgutils.cpp:2783]: "it's" -> "its" ?

File: test/compiler/codegen.jl [hunk 2]
```diff
@@ -938,3 +938,12 @@ BigStructAnyInt() = BigStructAnyInt((Union{Base.inferencebarrier(Float64), Int}=
 @test egal_any54109(Torture1_54109(), Torture1_54109())
 @test egal_any54109(Torture2_54109(), Torture2_54109())
 @test !egal_any54109(Torture1_54109(), Torture1_54109((DefaultOr54109(2.0, false) for i = 1:897)...))
+
+function foo54599()
+    pkgid = Base.identify_package("Test")
```
Comment (KristofferC) [inline test/compiler/codegen.jl:943]: Is it possible to minimise this a bit or is that tricky? It's a lot of code under the hood here and with changes to that code it's possible a regression would not be found?

**Comment by topolarity** [comment]
Uh oh, I found this example that fails LLVM assertions even with this PR:
```julia
bar54599() = Base.inferencebarrier(true) ? (Base.PkgId(Test),1) : nothing
function foo54599()
    pkginfo = @noinline bar54599()
    pkgid = pkginfo !== nothing ? pkginfo[1] : nothing
    pkgid.uuid !== nothing ? pkgid.uuid : false
end

@test foo54599() !== nothing
```
```
julia: /home/topolarity/repos/julia/src/llvm-alloc-opt.cpp:408: void {anonymous}::Optimizer::insertLifetime(llvm::Value*, llvm::Constant*, llvm::Instruction*): Assertion `cur.p_cur != cur.p_end' failed.
```

**Comment by gbaraldi** [comment]
@topolarity that is a separate crash 😄, that is fixed in https://github.com/JuliaLang/julia/pull/54690.

**Comment by oscardssmith** [comment]
in that case, should we merge this?

**Comment by gbaraldi** [comment]
I wasn't able to minimize that test to I gues so

**Comment by topolarity** [comment]
The crash I pointed out was in search of a more minimal test case.

Here's one that works:

```julia

bar54599() = Base.inferencebarrier(true) ? (Base.PkgId(Main),1) : nothing
function foo54599()
    pkginfo = @noinline bar54599()
    pkgid = pkginfo !== nothing ? pkginfo[1] : nothing
    @noinline println(devnull, pkgid)
    pkgid.uuid !== nothing ? pkgid.uuid : false
end
foo54599() # segfaults
```

**Pull request merged**